### ISSUE TITLE
Cargo in WSL prefers windows user config

### ISSUE THREAD

**Issue by weltensturm** [rust-lang/cargo#11677]
### Problem

Cargo in WSL uses the windows host cargo config in `%UserProfile%/.cargo/config.toml`, and it looks like values there are preferred over the WSL `~/.cargo/config.toml`.
When there is either `rustc-wrapper` or `target-dir` configured, cargo in WSL fails because it does not have access to those paths.

I have a `~/.cargo/config.toml` in WSL with the following content:
```
[build]
rustc-wrapper = "/home/<user>/.cargo/bin/sccache"
target-dir = "/home/<user>/.cargo/targets"
```
But cargo prefers the host values.
Error when `rustc-wrapper` is set on windows:
```
error: failed to run `rustc` to learn about target-specific information

Caused by:
  could not execute process `/mnt/c/Users/<user>/C:/Users/<user>/.cargo/bin/sccache.exe rustc - --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=cfg` (never executed)

Caused by:
  No such file or directory (os error 2)
```
Error when `target-dir` is set on windows:
```
error: failed to join paths from `$LD_LIBRARY_PATH` together

Check if any of path segments listed below contain an unterminated quote character or path separator:
    "/mnt/c/Users/<user>/C:/Users/bael/.cargo/targets/debug/deps"
    "/home/<user>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib"
    "/home/<user>/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib"

Caused by:
  path segment contains separator `:`
```

`which { rustup cargo rustc }` shows WSL binaries, so there is no issue with accidentally invoking host binaries.

When I delete the contents of the windows `cargo.toml`, the WSL config is applied correctly.

### Notes

`wsl -v -l`
```
WSL version: 1.0.3.0
Kernel version: 5.15.79.1
WSLg version: 1.0.47
MSRDC version: 1.2.3575
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.1105
```

### Version

```text
cargo 1.67.0 (8ecd4f20a 2023-01-10)
release: 1.67.0
commit-hash: 8ecd4f20a9efb626975ac18a016d480dc7183d9b
commit-date: 2023-01-10
host: x86_64-unknown-linux-gnu
libgit2: 1.5.0 (sys:0.16.0 vendored)
libcurl: 7.86.0-DEV (sys:0.4.59+curl-7.86.0 vendored ssl:OpenSSL/1.1.1q)
os: Ubuntu 22.04 (jammy) [64-bit]
```

**Comment by weihanglo** [comment]
My guess is that `home` crate only respects `USERPROFILE` even on WSL.

https://github.com/rust-lang/cargo/blob/82c3bb79e3a19a5164e33819ef81bfc2c984bc56/crates/home/src/windows.rs#L9-L30

I don't really have my Windows close by though. I believe @epage is more helpful. @epage, do you have any experience in WSL with home directory setup?

**Comment by weltensturm** [comment]
It works when copying the project to a "native" directory, it's only when the project is in some way under a symlink to `/mnt/c/Users/<user>/` or any child directory where the problem occurs. I assume it happens because rust unwraps the symlink path and uses the real/physical path to search all parent directories for configs.
I only checked that case after I created this issue, so apart from the edit block I left the description largely as it is, to show how confusing it is from the user side. But it's not as bad as it sounds, should I add a "when project is in a symlink to the windows user directory` to the title?

**Comment by arlosi** [comment]
>  I assume it happens because rust unwraps the symlink path and uses the real/physical path to search all parent directories for configs

Yes, this is correct (it's unrelated to WSL). `std::env::current_dir()` calls `getcwd`, on Linux, which resolves symlinks. Cargo never sees the version of the path that contains the symlinks.

**Comment by n-tropy247** [comment]
> My guess is that `home` crate only respects `USERPROFILE` even on WSL.
> 
> https://github.com/rust-lang/cargo/blob/82c3bb79e3a19a5164e33819ef81bfc2c984bc56/crates/home/src/windows.rs#L9-L30
> 
> I don't really have my Windows close by though. I believe @epage is more helpful. @epage, do you have any experience in WSL with home directory setup?

Can confirm, this resolved for me after creating a USERPROFILE environment variable in WSL that pointed to the right location.

**Comment by bingzhux** [comment]
this issue makes me struggled for a very very long time. 

Actually, all I wanted is just very simple:  
`As long as I enter WSL2 environment to build my rust project, cargo should always use WSL2's cargo configuration file, which is located at WSL2:/home/wsl2user/.cargo/config.toml`

but currently in WSL2:

- If my current rust project folder is under **/mnt/c/Users/winuser/...**, cargo will use host Windows configuration file, which is located at **Windows: C:\Users\winuser\\.cargo\config.toml**

- If my current rust project folder is under other non-mnt directories, e.g. **/home/wsl2user/rustdev/.**, cargo will use WSL2's cargo configuration file, which is **WSL2: /home/wsl2user/.cargo/config.toml**

**Comment by weihanglo** [comment]
Forgot to say, `home` was migrated into `rust-lang/cargo` repo <https://github.com/rust-lang/cargo/pull/11359> last year, so cargo maintainers here do have permissions to update it.

**Comment by CGMossa** [comment]
What's the status of this? I tried to add `.cargo/config.toml` with an `[env]`, but it won't detect those variables either. I'm detecting them via `env::var_os`.

**Comment by arlosi** [comment]
@CGMossa what your encountering sounds like a different problem. If you can create a minimal reproduceable example, please create a new issue with details.

**Comment by RReverser** [comment]
Running into this on a regular basis too. Setting USERPROFILE doesn't seem to work as a workaround, and copying projects each time I want to do a quick check that it still compiles on Linux is not very sustainable either. Is there any functioning workaround?
